Implements Contour Namespace API

danehans opened the pull request:
Implements the Namespace API added in https://github.com/projectcontour/contour-operator/pull/20. Updates `ensureNamespace()` and `ensureNamespaceRemoved()` to use a `Contour` object instead of a namespace string as a parameter. Updates `ensureNamespaceRemoved()` to only remove the namespace if RemoveOnDeletion is unset and the namespace does not match a name in namespaceBlacklist.

Requires PR https://github.com/projectcontour/contour-operator/pull/24

/assign @jpeach @stevesloka 
/cc @Miciah

Signed-off-by: Daneyon Hansen <daneyonhansen@gmail.com>

stevesloka commented on controller/contour/namespace.go:
```diff
@@ -20,20 +20,20 @@ import (
-	defaultContourNamespace = "projectcontour"
-)
+// namespaceBlacklist is a list of namespace names that should not be removed.
+var namespaceBlacklist = []string{"contour-operator", "default"}
```
Need a new name for this, maybe something like `coreNamespaceList` or something?

stevesloka replied:
Also should add `kube-system` to this list or have other logic to not allow specific namespaces to be created?

danehans replied:
@stevesloka thanks for the review. You're spot-on with both comments. I'll work on the updates and push a new commit.

danehans commented:
@stevesloka commit `aef797c` addresses your feedback and also implements a portion of the API that was previously missing (not removing the namespace if another contour exists), PTAL.

/cc @Miciah

skriss commented on controller/contour/controller.go:
```diff
@@ -103,26 +103,47 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
 
 // ensureContour ensures all necessary resources exist for the given contour
 // in namespace ns.
```
nit: the godoc comment still refers to `ns`.

danehans replied:
ack

skriss commented on controller/contour/controller.go:
```diff
@@ -103,26 +103,47 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
 }
 
 // ensureContourRemoved ensures all resources for the given contour do not exist
 // in namespace ns.
```
nit: the godoc comment still refers to `ns`.

danehans replied:
ack

skriss commented on controller/contour/namespace.go:
```diff
@@ -50,21 +50,45 @@ func (r *Reconciler) ensureNamespace(ctx context.Context, name string) error {
-		r.Log.Info("namespace does not exist; skipping removal", "name", ns.Name)
+// ensureNamespaceRemoved ensures the namespace for the provided contour is removed,
+// bypassing deletion if any of the following conditions apply:
+//   - RemoveOnDeletion is set.
```
should be "RemoveOnDeletion is **not** set", or "...set to false".

danehans replied:
ack

skriss commented on controller/contour/controller.go:
```diff
@@ -103,26 +103,47 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+// contoursExist lists Contour objects in all namespaces if all is true or only the
+// namespace of contour otherwise, returning true if any exist other than contour.
+func (r *Reconciler) contoursExist(ctx context.Context, contour *operatorv1alpha1.Contour, all bool) (bool, error) {
```
`all` is not currently specified as `true` anywhere, but I assume you have a plan to use it?

danehans replied:
I replaced this func with https://github.com/projectcontour/contour-operator/pull/21/files#diff-eee6f38bbe0411df081f7393625f732eR136-R159. I'll remove either based on which PR merges first.

skriss commented on controller/contour/controller.go:
```diff
@@ -103,26 +103,47 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+// contoursExist lists Contour objects in all namespaces if all is true or only the
+// namespace of contour otherwise, returning true if any exist other than contour.
+func (r *Reconciler) contoursExist(ctx context.Context, contour *operatorv1alpha1.Contour, all bool) (bool, error) {
```
maybe `otherContoursExist` or something like that would be slightly more clear.

danehans replied:
ack

skriss reviewed (changes requested):
handful of comments but this mostly looks good!

danehans commented:
@skriss I really appreciate the review. Commit `c3ac3d3` addresses your feedback, PTAL.

skriss commented:
Needs a rebase now that #19 has merged.

danehans commented:
@skriss commit `58ac6b6` resolves the conflict, PTAL.

skriss reviewed (approved):
LGTM :shipit:

danehans commented:
@stevesloka PTAL when you have a moment.
